engine/src/main/java/org/terasology/engine/core/module/ClasspathCompromisingModuleFactory.java

```diff
@@ -52,6 +63,97 @@ public Module createArchiveModule(ModuleMetadata metadata, File archive) throws
+                "metadataName `%s` is not in loader map", metadataName);
+        if (metadataUrl.getProtocol().equals("jar")) {
+            return modulePathFromMetadataJarUrl(metadataUrl);
+        } else {
+            return modulePathFromMetadataFileUrl(metadataName, metadataUrl);
+        }
+    }
+
+    private Path modulePathFromMetadataFileUrl(String metadata_name, URL url) {
+        Path path = fromUrl(url);
+        // We are considering the location of a resource file, so compare it to the code path.
+        // Include the metadata_name in case it has path components of its own.
+        Path relativePathFromModuleRoot = Paths.get(getDefaultCodeSubpath(), metadata_name);
+        return findModuleRoot(relativePathFromModuleRoot, path).orElse(path);
+    }
+
+    private Path modulePathFromMetadataJarUrl(URL url) {
+        checkArgument(url.getProtocol().equals("jar"), "Not a jar URL: %s", url);
+        try {
+            JarURLConnection connection = (JarURLConnection) url.openConnection();
+            url = connection.getJarFileURL();
+            // despite the method name, openConnection doesn't open anything unless we
+            // call connect(), so we needn't clean up anything here.
+        } catch (IOException e) {
+            throw new RuntimeException("Failed to get file from " + url, e);
+        }
+        Path path = fromUrl(url);
+        // We are considering the location of a jar file, so compare it to the libs path.
+        Path relativePathFromModuleRoot = Paths.get(getDefaultLibsSubpath());
+        // Assume jars would be directly in the libs path (not in a subdirectory).
+        Path jarDirectory = path.getParent();
+        return findModuleRoot(relativePathFromModuleRoot, jarDirectory).orElse(path);
+    }
+
+    /**
+     * Find the root of a module build directory.
+     * <p>
+     * If {@code path} matches a known build directory sub-path, return the base directory.
+     * <p>
+     * Example:
+     * <ul>
+     *   <li>findModuleRoot("foo/bar.txt", "/some/base/foo/bar.txt") == "/some/base"
+     *   <li>findModuleRoot("foo/bar.txt", "/some/base/x/foo/bar.txt") == "/some/base/x"
+     *   <li>findModuleRoot("foo/bar.txt", "/some/base/foo/x/bar.txt") == <i>Empty</i>
+     *   <li>findModuleRoot("foo/bar.txt", "/some/base/foo/baz.txt") == <i>Empty</i>
+     * </ul>
+     */
+    private static Optional<Path> findModuleRoot(Path relativePathFromModuleRoot, Path path) {
+        if (path.endsWith(relativePathFromModuleRoot)) {
+            int relativeDepth = relativePathFromModuleRoot.getNameCount();
+            Path parentPath = path.subpath(0, path.getNameCount() - relativeDepth);
+            if (path.getRoot() != null) {  // TODO: test case
+                parentPath = path.getRoot().resolve(parentPath);
+            }
+            return Optional.of(parentPath);
+        } else {
+            logger.warn(" +- does not seem to be in a build directory {}", path);
```

Comment:
Does the ` +-` prefix in this logging line make more sense in the context this method is called?

Reply:
The indentation helped when I had a bunch more logging earlier. but yeah, it's worth a second look to see if it still helps.